oauth2_http/java/com/google/auth/oauth2/CertificateIdentityPoolSubjectTokenSupplier.java

```diff
@@ -97,33 +119,157 @@ private static String encodeCert(X509Certificate certificate)
+        // Get the first certificate from the user-provided trust chain file.
+        X509Certificate firstTrustCert = trustChainCerts.get(0);
+        String encodedFirstTrustCert = encodeCert(firstTrustCert);
+
+        // If the first certificate in the user-provided trust chain file is *not* the leaf
+        // certificate (which has already been added as the first element to `certChain`), then add
+        // this certificate to `certChain`. This handles cases where the user's trust chain file
+        // starts with an intermediate certificate. If the first certificate in the trust chain file
+        // *is* the leaf certificate, this means the user has explicitly included the leaf in their
+        // trust chain file. In this case, we skip adding it again to prevent duplication, as the
+        // leaf is already at the beginning of `certChain`.
+        if (!encodedFirstTrustCert.equals(encodedLeafCert)) {
+          certChain.add(encodedFirstTrustCert);
+        }
+
+        // Iterate over the remaining certificates in the trust chain.
+        for (int i = 1; i < trustChainCerts.size(); i++) {
+          X509Certificate currentCert = trustChainCerts.get(i);
+          String encodedCurrentCert = encodeCert(currentCert);
 
-      java.util.List<String> certChain = new java.util.ArrayList<>();
-      certChain.add(encodedLeafCert);
+          // Throw an error if the current certificate (from the user-provided trust chain file,
+          // at an index beyond the first) is the same as the leaf certificate.
+          // This enforces that if the leaf certificate is included in the trust chain file by the
+          // user, it must be the very first certificate in that file. It should not appear
+          // elsewhere in the chain.
+          if (encodedCurrentCert.equals(encodedLeafCert)) {
+            throw new IllegalArgumentException(
+                "The leaf certificate should only appear at the beginning of the trust chain file, or be omitted entirely.");
+          }
 
-      return OAuth2Utils.JSON_FACTORY.toString(certChain);
+          // Add the current certificate to the chain.
+          certChain.add(encodedCurrentCert);
+        }
+      }
+    } catch (IllegalArgumentException e) {
+      // This catches the specific error for misconfigured trust chain (e.g., leaf in wrong place).
+      throw new IOException("Trust chain misconfiguration: " + e.getMessage(), e);
+    } catch (NoSuchFileException e) {
+      throw new IOException(String.format("Trust chain file not found: %s", trustChainPath), e);
     } catch (CertificateException e) {
-      // Catch CertificateException to provide a more specific error message including
-      // the path of the file that failed to parse, and re-throw as IOException
-      // as expected by the getSubjectToken method signature for I/O related issues.
+      // This catches parsing errors if certificates in the trust chain file are invalid.
```

Comment:
nit: this comment seems a bit redundant, the exception message seems self-explainatory.